pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -178,55 +227,73 @@ func (p *ecrProvider) Enabled() bool {
-func (p *ecrProvider) Provide() credentialprovider.DockerConfig {
-	cfg := credentialprovider.DockerConfig{}
+// GetAuthorizationToken gets the ECR authorization token using the ECR API
+func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenInput) (*ecr.GetAuthorizationTokenOutput, error) {
+	return p.svc.GetAuthorizationToken(input)
+}
 
-	// TODO: fill in RegistryIds?
-	params := &ecr.GetAuthorizationTokenInput{}
-	output, err := p.getter.GetAuthorizationToken(params)
+type cacheEntry struct {
+	expiresAt   time.Time
+	credentials credentialprovider.DockerConfigEntry
+	registry    string
+}
+
+// makeCacheEntry decodes the ECR authorization entry and re-packages it into a
+// cacheEntry.
+func makeCacheEntry(data *ecr.AuthorizationData, registry string) (*cacheEntry, error) {
+	decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
 	if err != nil {
-		klog.Errorf("while requesting ECR authorization token %v", err)
-		return cfg
+		return nil, err
 	}
-	if output == nil {
-		klog.Errorf("Got back no ECR token")
-		return cfg
+	parts := strings.SplitN(string(decodedToken), ":", 2)
+	creds := credentialprovider.DockerConfigEntry{
+		Username: parts[0],
+		Password: parts[1],
+		Email:    "not@val.id", // ECR doesn't care and Docker is about to obsolete it
 	}
+	return &cacheEntry{
+		expiresAt:   *data.ExpiresAt,
+		credentials: creds,
+		registry:    registry,
+	}, nil
+}
 
-	for _, data := range output.AuthorizationData {
-		if data.ProxyEndpoint != nil &&
-			data.AuthorizationToken != nil {
-			decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
-			if err != nil {
-				klog.Errorf("while decoding token for endpoint %v %v", data.ProxyEndpoint, err)
-				return cfg
-			}
-			parts := strings.SplitN(string(decodedToken), ":", 2)
-			user := parts[0]
-			password := parts[1]
-			entry := credentialprovider.DockerConfigEntry{
-				Username: user,
-				Password: password,
-				// ECR doesn't care and Docker is about to obsolete it
-				Email: "not@val.id",
-			}
-
-			klog.V(3).Infof("Adding credentials for user %s in %s", user, p.region)
-			// Add our config entry for this region's registry URLs
-			cfg[p.regionURL] = entry
-
-		}
-	}
-	return cfg
+// ecrExpirationPolicy implements ExpirationPolicy from client-go.
+type ecrExpirationPolicy struct{}
+
+// stringKeyFunc returns the cache key as a string
+func stringKeyFunc(obj interface{}) (string, error) {
+	key := obj.(*cacheEntry).registry
+	return key, nil
+}
+
+// IsExpired checks if the ECR credentials are expired. Attempt to renew
+// credentials at their half-life to avoid using potentially expired
+// credentials.
+func (p *ecrExpirationPolicy) IsExpired(entry *cache.TimestampedEntry) bool {
+	ecrExpiresAt := entry.Obj.(*cacheEntry).expiresAt
+	halfLife := ecrExpiresAt.Sub(entry.Timestamp) / time.Duration(2)
+	return ecrExpiresAt.Add(halfLife * -1).Before(time.Now())
```

Comment:
I'm fine with this, but it is a big change from expiring 5 minutes before - I know in the past we had some push-back over doubling the number of ECR token requests.

The other option is to store the actual expiration time in `cacheEntry`, rather than the ECR expiration time.  I think that then means you also don't need to look at `entry.Timestamp`, so one less dependency there.

Reply:
Will poke at this one tomorrow. Sleep calls.

Reply:
Updated!